turbopack/crates/turbopack-ecmascript/src/analyzer/graph.rs

```diff
@@ -1566,101 +1624,78 @@ impl VisitAstPath for Analyzer<'_> {
         decl: &'ast FnDecl,
         ast_path: &mut AstNodePath<AstParentNodeRef<'r>>,
     ) {
-        let old = self
-            .cur_fn_return_values
-            .replace(get_fn_init_return_vals(decl.function.body.as_ref()));
-        let old_ident = self.cur_fn_ident;
-        self.cur_fn_ident = decl.function.span.lo.0;
-        decl.visit_children_with_ast_path(self, ast_path);
-        let return_value = self.take_return_values();
+        let fn_value = self.enter_fn(decl.function.span.lo.0, false, |this| {
+            decl.visit_children_with_ast_path(this, ast_path);
+        });
+        // Why do we do this? Move all effects to hoisted effects after visiting a function
+        // declaration?
         self.hoisted_effects.append(&mut self.effects);
-
-        self.add_value(
-            decl.ident.to_id(),
-            JsValue::function(self.cur_fn_ident, return_value),
-        );
-
-        self.cur_fn_ident = old_ident;
-        self.cur_fn_return_values = old;
+        self.add_value(decl.ident.to_id(), fn_value);
     }
 
     fn visit_fn_expr<'ast: 'r, 'r>(
         &mut self,
         expr: &'ast FnExpr,
         ast_path: &mut AstNodePath<AstParentNodeRef<'r>>,
     ) {
-        let old = self
-            .cur_fn_return_values
-            .replace(get_fn_init_return_vals(expr.function.body.as_ref()));
-        let old_ident = self.cur_fn_ident;
-        self.cur_fn_ident = expr.function.span.lo.0;
-        expr.visit_children_with_ast_path(self, ast_path);
-        let return_value = self.take_return_values();
-
+        let fn_value = self.enter_fn(expr.function.span.lo.0, false, |this| {
+            expr.visit_children_with_ast_path(this, ast_path);
+        });
+        // function expressions may have an ident when they appear in an export declaration.
```

Comment:
Function expressions can always have an ident? It's just usually not used for anything other than stack traces or react component names.

![Screenshot 2025-08-21 at 2.28.52 PM.png](https://app.graphite.dev/user-attachments/assets/638ee303-0ef4-4236-bef3-53a7994a11d0.png)

Reply:
Ah thanks, i got mixed up i guess by some of the docs on exports. which say that in  `export default function foo(){}` `foo` is a function expression even though it looks like a declaration